apps/query-server/queryserver/main.py

```diff
@@ -67,29 +97,158 @@ async def get_index(
+                "data": "Generating plan",
+            }
+            await asyncio.sleep(0.1)
+            plan = sqclient.generate_plan(query.query, query.index, sqclient.get_opensearch_schema(query.index))
+            logger.info(f"Generated plan: {plan}")
+            # Don't want to return these through the API.
+            plan.llm_plan = None
+            plan.llm_prompt = None
+            yield {
+                "event": "plan",
+                "data": plan.model_dump_json(),
+            }
+            await asyncio.sleep(0.1)
+            logger.info("Running plan")
+            yield {
+                "event": "status",
+                "data": "Running plan",
+            }
+            await asyncio.sleep(0.1)
+            sqresult = sqclient.run_plan(plan)
+            t1 = time.time()
+            num_results = 0
+            if isinstance(sqresult.result, DocSet):
+                logger.info("Got DocSet result")
+                for doc in sqresult.result.take_all():
+                    rendered = doc_to_json(doc)
+                    logger.info(f"Doc: {rendered}")
+                    if rendered is not None:
+                        num_results += 1
+                        yield {
+                            "event": "result_doc",
+                            "data": rendered,
+                        }
+                        await asyncio.sleep(0.1)
+            else:
+                num_results += 1
+                yield {
+                    "event": "result",
+                    "data": sqresult.result,
+                }
+                await asyncio.sleep(0.1)
+
+            for doc in sqresult.retrieved_docs():
+                yield {
+                    "event": "retrieved_doc",
+                    "data": doc,
+                }
+                await asyncio.sleep(0.1)
+
+            t2 = time.time()
+            logger.info(f"Finished query in {t2 - t1:.2f} seconds with {num_results} results")
+            yield {
+                "event": "status",
+                "data": f"Query complete - {num_results} results in {t2 - t1:.2f} seconds",
+            }
+            await asyncio.sleep(0.1)
+        except asyncio.CancelledError:
+            logger.info("Disconnected from client")
+
+    return EventSourceResponse(query_runner())
+
+
+@app.post("/v1/query", response_model=None)
+async def run_query(query: Query) -> Union[EventSourceResponse, QueryResult]:
+    """Run the given query.
+
+    If the `stream` parameter is set to true, the result will be streamed back to the client as a series of SSE events.
+    Otherwise, the result will be returned as a QueryResult object.
+    """
+
+    logger.info(f"Running query: {query}")
+
+    if query.query is None and query.plan is None:
+        raise ValueError("query or plan is required")
+    if query.query is not None and query.plan is not None:
+        raise ValueError("query and plan cannot both be specified")
+
+    if query.stream:
```

Comment:
I don't understand why we wouldn't always stream results (once we fix it to actually stream something).

However, if we don't want to, I would have thought this would become:
```
ret = []
for i in run_query_stream(query):
    ret.append(i)
return ret.join("")
```

Reply:
Because the output formats are different in a streaming and non-streaming case. I don't want to give them SSE format in a non-streaming instance, just JSON. This is going to be a lot easier for them to consume.

Reply:
Seems like the intent here is unclear. For at least the platform demo we need this API to be streaming, so query execution is not blocking things like returning the generated plan, and status updates from the server while the query is running. I expect our customers will want this as well, but I made streaming optional in case they just want to get back a single JSON response. The formats of the streamed and non-streamed responses are fundamentally different, so one cannot just do as you say and concatenate them together.

For this I am using SSE which is a "standard" in the sense that browsers support it natively, and OpenAI (and many others) use it to stream partial responses on a REST API. Unfortunately the docs on the web on SSE are kind of crappy but this page explains the protocol reasonably well: https://hpbn.co/server-sent-events-sse/#event-stream-protocol

However if one is not streaming, just returning JSON makes the most sense.

Reply:
Ok. I'm not a fan of having two separate implementations of query execution, you have to fix both of them when you make changes.   For expediency it's worth getting in this way, but I'd recommend writing something that can combine the events into the json response. Something like:
```
ret = {}
for e in run_query_streamed():
   if e["event"] == "status":
       continue # drop, we're already done
   ret[e["event"]] = e["data"]
```
would mean  you have one implementation of run_query and one transform of events into a json return value. You could also choose at this point to stream back the single json event with a status block like the model server does, or do smarter combining of events, e.g. turn something into an array.


FWIW, I found the missing assertion that query was not none in the async version by comparing with the sync one.  There were also a bunch of differences in logging.

Reply:
For now I'd like the plain JSON result to be very close to SycamoreQueryResult. I can refactor this later.